src/aggregate.d

```diff
@@ -215,78 +208,136 @@ public:
             sd.semanticTypeInfoMembers();
     }
 
-    abstract void finalizeSize();
-
-    override final uint size(Loc loc)
+    /***************************************
+     * Find all instance fields, then push them into `fields`.
+     *
+     * Runs semantic() for all instance field variables, but also
+     * the field types can reamin yet not resolved forward references,
+     * except direct recursive definitions.
+     * After the process sizeok is set to SIZEOKfwd.
+     *
+     * Returns:
+     *      false if any errors occur.
+     */
+    final bool determineFields()
     {
-        //printf("AggregateDeclaration::size() %s, scope = %p\n", toChars(), scope);
-        if (loc.linnum == 0)
-            loc = this.loc;
-        if (sizeok != SIZEOKdone && _scope)
-        {
-            semantic(null);
+        if (sizeok != SIZEOKnone)
+            return true;
 
-            // Determine the instance size of base class first.
-            if (ClassDeclaration cd = isClassDeclaration())
-                cd.baseClass.size(loc);
-        }
+        //printf("determineFields() %s, fields.dim = %d\n", toChars(), fields.dim);
 
-        if (sizeok != SIZEOKdone && members)
+        extern (C++) static int func(Dsymbol s, void* param)
```

Comment:
OT: I think in the near future we should change `Dsymbol.apply` as it can take a D delegate. After that, the arbitrary named local function can be replaced with unnamed lambda.